### PR TITLE
Fixed #30557 -- Fixed crash of ordering when parent model's Meta.ordering contains expressions.

### PR DESCRIPTION
**Pull request by ekeydar**
Add check for instance of OrderBy inside got_order_dir. Not sure if this is full solution, but it solves the bugs in the ticket test cases.

### CODE REVIEW

File: tests/model_inheritance/models.py [hunk 1]
```diff
@@ -91,6 +91,9 @@ class Place(models.Model):
     def __str__(self):
         return "%s the place" % self.name
 
+    class Meta:
+        ordering = [models.F('name').desc(nulls_last=True)]
+
```
Comment (felixxm) [inline tests/model_inheritance/models.py]: This change is not related with test and can be reverted.

File: django/db/models/sql/query.py [hunk 2]
```diff
@@ -2198,6 +2198,9 @@ def get_order_dir(field, default='ASC'):
     prefix) should sort. The '-' prefix always sorts the opposite way.
     """
     dirn = ORDER_DIR[default]
+    if isinstance(field, OrderBy):
+        return field.expression.name, 'DESC' if field.descending else 'ASC'
+
```
Comment (felixxm) [inline django/db/models/sql/query.py]: I don't think that it is a proper solution because `get_order_dir()` returns only field name and direction, so it will not properly in case of using database functions in `meta.ordering`. I think we should fix this in `find_ordering_name()`, e.g.
```diff
diff --git a/django/db/models/sql/compiler.py b/django/db/models/sql/compiler.py
index a44adfc760..fb2a4b23b1 100644
--- a/django/db/models/sql/compiler.py
+++ b/django/db/models/sql/compiler.py
@@ -716,6 +716,9 @@ class SQLCompiler:
 
             results = []
             for item in opts.ordering:
+                if isinstance(item, OrderBy):
+                    results.append((item, False))
+                    continue
                 results.extend(self.find_ordering_name(item, opts, alias,
                                                        order, already_seen))
             return results
```

**Comment by cansarigol** [comment]
Hi @ekeydar , you should check F instance (ordering = [models.F('name')]) I was working on this issue, I focused this part if it can help you. 

+++ b/django/db/models/sql/compiler.py
@@ -716,8 +716,12 @@ class SQLCompiler:
 
```
             results = []
             for item in opts.ordering:
-                results.extend(self.find_ordering_name(item, opts, alias,
-                                                       order, already_seen))
+                if isinstance(item, OrderBy):
+                    results.extend([(item, False)])
+                elif isinstance(item, F):
+                    results.extend([(OrderBy(item, descending=descending), False)])
+                else:
+                    results.extend(self.find_ordering_name(item, opts, alias, order, already_seen))
             return results
         targets, alias, _ = self.query.trim_joins(targets, joins, path)
         return [(OrderBy(transform_function(t, alias), descending=descending), False) for t in targets]
```

File: tests/ordering/tests.py [hunk 3]
```diff
@@ -462,6 +462,12 @@ def test_default_ordering_by_f_expression(self):
             attrgetter('headline')
         )
 
+    def test_default_parent_ordering_by_expression(self):
+        sub1 = SubArticle.objects.create(headline='Sub 1', pub_date=datetime(2005, 7, 26))
+        sub2 = SubArticle.objects.create(headline='Sub 2', pub_date=datetime(2005, 7, 27))
+        articles = SubArticle.objects.order_by('article_ptr')
+        self.assertSequenceEqual(articles, [sub2, sub1])
```
Comment (felixxm) [inline tests/ordering/tests.py]: We should add third `SubArtictle` to check if expressions are used in ordering, e.g.
```diff
diff --git a/tests/ordering/tests.py b/tests/ordering/tests.py
index 39b5a75564..886c97e60f 100644
--- a/tests/ordering/tests.py
+++ b/tests/ordering/tests.py
@@ -464,9 +464,10 @@ class OrderingTests(TestCase):
 
     def test_default_parent_ordering_by_expression(self):
         sub1 = SubArticle.objects.create(headline='Sub 1', pub_date=datetime(2005, 7, 26))
-        sub2 = SubArticle.objects.create(headline='Sub 2', pub_date=datetime(2005, 7, 27))
+        sub2 = SubArticle.objects.create(headline='Sub 2', pub_date=datetime(2005, 7, 27), author=self.author_2)
+        sub3 = SubArticle.objects.create(headline='Sub 2', pub_date=datetime(2005, 7, 27), author=self.author_1)
         articles = SubArticle.objects.order_by('article_ptr')
-        self.assertSequenceEqual(articles, [sub2, sub1])
+        self.assertSequenceEqual(articles, [sub3, sub2, sub1])
```

**Comment by felixxm** [comment]
Updated in #11555.
